library/cwm/src/lib/cwm/widget.rb

```diff
@@ -0,0 +1,425 @@
+    end
+
+    def cleanup_method
+      fun_ref(method(:cleanup), "void (string)")
+    end
+
+    def validate_method
+      fun_ref(method(:validate), "boolean (string, map)")
+    end
+  end
+
+  class CustomWidget < AbstractWidget
+    def description
+      unless respond_to?(:content)
+        raise "For custom widget '#{self.class}' content method have to be defined"
+      end
+
+      {
+        "custom_widget" => content,
+        "widget"        => :custom
+      }.merge(super)
+    end
+  end
+
+  class EmptyWidget < AbstractWidget
+    def initialize(id)
+      self.widget_id = id
+    end
+
+    def description
+      {
+        "widget"        => :empty
+      }.merge(super)
+    end
+  end
+
+  # helpers for easier set/obtain value of widget for widgets where value is
+  # obtained by :Value symbol
+  module ValueBasedWidget
+
+  protected
+
+    def value
+      Yast::UI.QueryWidget(Id(widget_id), :Value)
+    end
+
+    def value=(val)
+      Yast::UI.ChangeWidget(Id(widget_id), :Value, val)
+    end
+  end
+
+  class InputFieldWidget < AbstractWidget
+    include ValueBasedWidget
+
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      {
+        "widget"        => :inputfield
+      }.merge(super)
+    end
+  end
+
+  class PasswordWidget < AbstractWidget
+    include ValueBasedWidget
+
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      {
+        "widget"        => :password
+      }.merge(super)
+    end
+  end
+
+  class CheckboxWidget < AbstractWidget
+    include ValueBasedWidget
+
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      {
+        "widget"        => :checkbox
+      }.merge(super)
+    end
+
+  protected
+
+    def checked?
+      value
+    end
+
+    def unchecked?
+      !value
+    end
+
+    def check
+      self.value = true
+    end
+
+    def uncheck
+      self.value = false
+    end
+  end
+
+  class ComboBoxWidget < AbstractWidget
+    include ValueBasedWidget
+
+    # description for combobox additionally support `items` method.
+    # `items` method have to return array of string pairs, where first value is
+    # item id and second is item label.
+    # @example items method
+    #   def items
+    #     [
+    #       [ "Canada", _("Canada")],
+    #       [ "USA", _("United States of America")],
+    #       [ "North Pole", _("Really cold place")],
+    #     ]
+    #   end
+    #
+    def description
+      unless respond_to?(:label)
+        raise "For input field widget '#{self.class}' label method have to be defined"
+      end
+
+      res = {
+        "widget"        => :combobox,
+      }
+      res["items"] = items if respond_to?(:items)
```

Comment:
Does this have the same behavior as defining `ComboBoxWidget#items` returning an empty list? Then that method would be a much better place to document it.